### PR TITLE
[DNM] controller, Refactor controller reconciler for updates

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
This PR introduces a change to the controller: MACs added/deleted due to an update change will be now handled in the controller and configMap. 
This fix tackles the current inability to handle webhook rejections in KMP that has been reproduced on the tier1 CNAO [test_id:2995](https://github.com/kubevirt/cluster-network-addons-operator/issues/720) which [fails](https://github.com/kubevirt/cluster-network-addons-operator/issues/720) occasionally.
The solution includes a change in the KMP controller design, as well as changes added to the tests to allow cache restoration in case of conflict, and tests meant to make sure the change works well.

- **vmToMacPoolMap, Add a vm name to interface:mac map**
In order to follow the macs inserted/deleted from any vm we add a new
map to the macPool module.
The function implemented to update this map is updateVmToMacPoolMap().
We update added macs in the webhook context to avoid allocation races,
and update removed macs when vm is deleted.
We also populate it during the init of the KMP manager
Added vmToMacPoolMap to vm macpool module
Added updateVmToMacPoolMap() to handle mac additions/removals

- **MarkVMAsReady, Change mac status according to vmToMacPoolMap**
Currently the mac status is updated only for newly created vms. The mac
status is updated according to the current vm spec retreived at the
moment of the controller runtime reconcile.
To avoid possible races between webhook and reconcile runs, we change
the controller reconciler to update the mac's status according to values
set in vmToMacPoolMap, which adds new macs during the webhook context.
Furthermore, we expand the controller recocile to also handle
updated/deleted macs. Macs added will be updated in the webhook context,
where deleted macs will be marked for deletion and removed only in
controller reconcile context.

-  **vmToMacPoolMap, Avoid rejecting a mac if a vm already has it**
Added a check in allocateRequestedVirtualMachineInterfaceMac when adding
a predefined mac to a vm, and the mac is already in the mac map,
to avoid rejecting it if it is already set in the vm according to
vmToMacPoolMap.

- **vmWaitingConfigMap, Support deleted macs**
In order to support mac update (addition/removal) restorations due to
webhook rejection, we also populate the vmWaiting ConfigMap with macs
marked for removal/update in the webhook context.
For that we expand the ConfigMap Entry to also hold the vm's namespaced
name, related interface, and current mac status (pending
deletion/addition)

- **webhook, Add logs to track conflicts in vm handler**

- **ci, Allow for random conflict handling in tests**
Occationally a webhook update request could get rejected due to
conlflict. These case are handled by design so that after a period of
time the rejected change is reverted.
Since our tests are also subjected to these conflicts, we should allow
for the kubemacpool cache to restore itself before going on with the
test.
Added retryOnConflictAndWaitForRestoreCache method that waits for the
configured waitTime should an update conflict occurr.
Changed all tests that update the vms to use this instead of the regular
retryOnConflict.
Added additional timeout to functest to allow for such conflict to be
handled.

- **ci, Add tests to check mac allocation/deletions**
Added tests to check mac updates: mac additions and removals.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by rhrazdil — CHANGES_REQUESTED** [review]
A few comments, nothing useful to fix the issue though

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:104]: you're right

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:102]: `vmName` is the namespaced: namespace/vmName

Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:102]: I agree

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:483]: if I create `releases` in the  `RetryOnConflict`context, then it will be lost outside of that scope.

Because the configmap is scanned according to the `VmToMacPoolMap` state, if I update  `VmToMacPoolMap` in the  `RetryOnConflict` context and the updates fails on conflict, then the second try will not be consistent..

For that reason, I want the `VmToMacPoolMap` changes to persist only if the configMap is successfully updated, and that releases will be cleaned on each retry...

File: tests/virtual_machines_test.go
Comment (RamLavi) [reply tests/virtual_machines_test.go:402]: thanks!

File: tests/virtual_machines_test.go
Comment (RamLavi) [reply tests/virtual_machines_test.go:510]: thanks!

File: pkg/pool-manager/virtualmachine_pool.go
Comment (rhrazdil) [reply pkg/pool-manager/virtualmachine_pool.go:483]: Oh of course! Missed it

**Comment by RamLavi** [comment]
[rebased](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/511de1a8546bca3b424577eeac0b72c82d3afb1f..169befd692d6a3ea81fb5a82a389e3f30b55a4cd) on top of #261 since the test are failing due to opt-out issue.

/hold until @261 is merged

**Comment by RamLavi** [comment]
/hold cancel 

#261 is merged

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [inline pkg/pool-manager/virtualmachine_pool.go:267]: perhaps I should check it sooner, using `vmToMacPoolMap` to look for allocated macs instead of the previos vm spec.
@qinqon what do you think?

**Comment by RamLavi** [comment]
/test pull-kubemacpool-e2e-k8s
run 1

File: pkg/pool-manager/virtualmachine_pool.go
Comment (qinqon) [inline pkg/pool-manager/virtualmachine_pool.go:450]: Let's pass `macAddress` directly to `createConfigMapEntry` so the transformation into ":", "-" is done there I would say that also the `time.Now().Format(time.RFC3339)` can be done there.

File: pkg/pool-manager/virtualmachine_pool.go
Comment (qinqon) [inline pkg/pool-manager/virtualmachine_pool.go:459]: ditto

**Review by qinqon — CHANGES_REQUESTED** [review]
First pass, isn't the configmap data change going to break upgrade ?

File: pkg/pool-manager/virtualmachine_pool.go
Comment (rhrazdil) [inline pkg/pool-manager/virtualmachine_pool.go:223]: As @qinqon mentions below, the `deltaInterfacesMap` probably contains also the content of `releaseOldAllocations`

File: pkg/pool-manager/virtualmachine_pool.go
Comment (rhrazdil) [inline pkg/pool-manager/virtualmachine_pool.go:581]: Maybe a dumb question, but why do we keep mac addresses with dashes in the configMap? I mean why not use only colons?

Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:581]: I think it will break the configmap yaml format. frankly I never tried it change it.
But it's in my bucket list for refactor changes, after this PR is done.

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:64]: `AllocateVirtualMachineMac` is only called when creating a vm. no need to release macs in this method.

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:107]: Done

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:120]: Done

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:140]: you're right. Done

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:100]: Done

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:223]: yes I see what you mean, it's probably safer to deep copy

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:670]: right. thanks!

File: pkg/pool-manager/pool.go
Comment (RamLavi) [reply pkg/pool-manager/pool.go:87]: they are not re-calculated, but rather the init uses the configmap in order to restore the cache (only the part that wasn't already removed as part of the controller runtime, for example if the webhook creating the vm was rejected). 

Checking CNAO upgrade - I noticed the configmap Data remains, so I agree we need to keep backwars compatibility, alongside with the new behavior. 

Do you agree?

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:450]: but macAddress is only part of the configmap's key, not value

totally agree about the TimeStamp.. Done that.

File: pkg/pool-manager/virtualmachine_pool.go
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:459]: Done the TimeStamp

File: tests/virtual_machines_test.go
Comment (RamLavi) [reply tests/virtual_machines_test.go:489]: there is already a polarion test (test_id: 2995) doing that.

Comment (RamLavi) [reply tests/virtual_machines_test.go:489]: I don't think so. it would basically be us checking k8s conflict mechanism
This test expects the request to fail, if this ever breaks we'd be able to follow the `By` steps and understand it with out issuing another test

**Comment by RamLavi** [comment]
@qinqon fixes [here](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/4df30a4699c4722b91a5ee9876e04579776e8d7e..d5d3e1cc2211918eaaa7aef47dc92237c2ad173b)

**Comment by RamLavi** [comment]
/hold
proposing a new, better solution.
holding off merging this change

**Comment by qinqon** [comment]
@RamLavi you can close this PR, but without removing the branch.

**Pull request closed without merging**
